subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileToolchainIntegrationTest.groovy

```diff
@@ -391,37 +608,27 @@ public class Foo {
             }
         """
 
-        file("src/main/java/Foo.java") << "public class Foo {}"
-
         when:
-        runWithToolchainConfigured(jdk)
+        withInstallations(jdk).run(":compileJava", "--info")
 
         then:
         outputDoesNotContain("Compiling with Java command line compiler")
         outputContains("Compiling with toolchain '${jdk.javaHome.absolutePath}'.")
         javaClassFile("Foo.class").exists()
+        jdk.javaVersion == JavaVersion.forClass(javaClassFile("Foo.class").bytes)
 
         where:
-        javaVersion << [
-            JavaVersion.VERSION_1_8,
-            JavaVersion.VERSION_1_9,
-            JavaVersion.VERSION_11,
-            JavaVersion.VERSION_12,
-            JavaVersion.VERSION_13,
-            JavaVersion.VERSION_14,
-            JavaVersion.VERSION_15,
-            JavaVersion.VERSION_16,
-            JavaVersion.VERSION_17
-        ]
+        // Range of Java versions from 8 up to the current
+        javaVersion << JavaVersion.values().findAll { it.isJava8Compatible() && JavaVersion.current().isCompatibleWith(it) }
```

Comment:
```suggestion
        javaVersion << JavaVersion.values().findAll { it.isJava8Compatible() }
```

Why the second condition? We want to compile on a different JDK which can be a version newer than the current one, right?

Reply:
I had a wrong impression that values like `JavaVersion.VERSION_20` or `JavaVersion.VERSION_HIGHER` would break the test. They do not, because of the `assumeNonNull` call for the resolved toolchain installation.

It's still a large area to figure out how to make sure these sort of tests are not silently skipped for most/all parameter combinations. This test in particular could be a tricky one to handle.

Following your suggestion, I fixed this in one of the additional commits.